### PR TITLE
[WIP] SALTO-1022 Add element list command

### PR DESCRIPTION
**Pull request by netama**
WIP - NOT READY FOR REVIEW.

Adding the `element list` exploration command for identifying cross-element dependencies based on reference expressions. 
The first part is a list of referenced elements that can be easily edited and transferred to other commands (for example using `xargs`). For now this is based only on the selected environment.

Still missing:
* Support for Netsuite references (will add before merging)
* Integration with the new elem id selectors (either before or after merging depending on when the selectors are ready)

Sample output (not final):

Option 1 (graph traversal):
Duplicates between different branches are intentional in this presentation (see ticket for more details).
```
$ salto element list salesforce.Account salesforce.Profile.instance.Admin.layoutAssignments.Asset_Asset_Layout --max-depth 2
✔ Finished loading workspace for environment env1
salesforce.Account
   salesforce.Contract
      salesforce.Contact
      salesforce.Pricebook2
      salesforce.StandardValueSet.instance.ContractStatus
      salesforce.User
   salesforce.DandBCompany
      salesforce.User
   salesforce.OperatingHours
      salesforce.User
   salesforce.StandardValueSet.instance.AccountOwnership
   salesforce.StandardValueSet.instance.AccountRating
   salesforce.StandardValueSet.instance.AccountType
   salesforce.StandardValueSet.instance.Industry
   salesforce.StandardValueSet.instance.LeadSource
   salesforce.User
      salesforce.CallCenter
      salesforce.Contact
      salesforce.Group
      salesforce.Individual
      salesforce.Profile

salesforce.Profile.instance.Admin.layoutAssignments.Asset_Asset_Layout
   salesforce.Layout.instance.Asset_Layout
      salesforce.Asset
      salesforce.Asset.field.AccountId
      salesforce.Asset.field.ContactId
      salesforce.Asset.field.CreatedById
      salesforce.Asset.field.Description
      salesforce.Asset.field.InstallDate
      salesforce.Asset.field.IsCompetitorProduct
      salesforce.Asset.field.LastModifiedById
      salesforce.Asset.field.Name
      salesforce.Asset.field.Price
      salesforce.Asset.field.Product2Id
      salesforce.Asset.field.ProductCode
      salesforce.Asset.field.PurchaseDate
      salesforce.Asset.field.Quantity
      salesforce.Asset.field.SerialNumber
      salesforce.Asset.field.Status
      salesforce.Asset.field.UsageEndDate
```

Option 2 (flat list):
```
$ salto element list salesforce.Account salesforce.Profile.instance.Admin.layoutAssignments.Asset_Asset_Layout --max-depth 2
✔ Finished loading workspace for environment env1
Listing dependencies in env1 for the specified elements
  salesforce.Asset
  salesforce.CallCenter
  salesforce.Contact
  salesforce.Contract
  salesforce.DandBCompany
  salesforce.Group
  salesforce.Individual
  salesforce.Layout.instance.Asset_Layout
  salesforce.OperatingHours
  salesforce.Pricebook2
  salesforce.Profile
  salesforce.StandardValueSet.instance.AccountOwnership
  salesforce.StandardValueSet.instance.AccountRating
  salesforce.StandardValueSet.instance.AccountType
  salesforce.StandardValueSet.instance.ContractStatus
  salesforce.StandardValueSet.instance.Industry
  salesforce.StandardValueSet.instance.LeadSource
  salesforce.User
```

### CODE REVIEW

**Comment by omrilit** [comment]
Currently netsuite has only references between instances (and their inner values).
I don't think that it requires a special netsuite handling here.
IMO if the reference is to an instance or its inner value we should extract the `topLevelParent` from the reference and use its elemID.

**Comment by omrilit** [comment]
How do we take care for duplications?
if a certain element has 2 references to the same instance would we see it twice?
if 2 different elements have reference to the same elementA , will the CLI display all of the inner references from elementA twice or we will identify it and will display it only once?

**Comment by netama** [comment]
@omrilit that's what I meant by handling the netsuite references :) I think we agreed on that part, it's just not covered in the implementation yet.
I added details on the ticket for the questions about handling duplicates, probably easier to discuss there.

File: packages/core/src/core/list.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        transformFunc: findReferences,
+      })
+      nextLevel.push(...deps)
+    })
+    elemRefs = uniq(nextLevel).filter(dep => !referencedElemIDs.has(dep.elemId.getFullName()))
+  }
+
+  return [...referencedElemIDs]
+    .filter(idStr => !hasReferencedAncestor(idStr))
+    .filter(elemID => !initialElemIDs.has(elemID))
```
Comment (netama) [inline packages/core/src/core/list.ts:125]: I wasn't sure if we should include the original list of ids or not in the flat list format (and if yes - if it should be in its own section or sorted together with the rest). Any preference? @tomermevorach

Comment (ori-moisis) [reply packages/core/src/core/list.ts:125]: considering the use case of using the output for `move-to-common` and the possibility that the input to the `list ` command might be a selector with wildcards, I think the input ids should be in the output.
personally I don't see a reason to separate them

**Comment by netama** [comment]
replaced by https://github.com/salto-io/salto/pull/1609

**Pull request closed without merging**